sql/core/src/main/scala/org/apache/spark/sql/catalyst/optimizer/InjectRuntimeFilter.scala

```diff
@@ -146,7 +147,7 @@ object InjectRuntimeFilter extends Rule[LogicalPlan] with PredicateHelper with J
           predicateReference ++ condition.references,
           hasHitFilter = true,
           hasHitSelectiveFilter = hasHitSelectiveFilter || isLikelySelective(condition))
-      case _: LeafNode => hasHitSelectiveFilter
+      case l: LeafNode => hasHitSelectiveFilter || hasSelectivePredicate(l)
```

Comment:
how about we add a new trait `LeafNodeWithAccurateStats`? `InMemoryRelation` will be the only implementation. Then here we can just do
```
case l: LeafNodeWithAccurateStats => l.stats.rowCount.exists(_ < a_threshold) 
```

Reply:
how about this ? then we can use accurate stats of `InMemoryRelation` after materialized.
```scala
trait LeafNodeWithAccurateStats extends LeafNode {
  def statsAvailable: Boolean
}
```